From 2f656fd1fccb2ff886319fda8e279655bc57fbe1 Mon Sep 17 00:00:00 2001
From: Potin Lai <potin.lai@quantatw.com>
Date: Wed, 27 Nov 2024 19:19:03 +0800
Subject: [PATCH 1/1] Revert "net/ncsi: change from ndo_set_mac_address to
 dev_set_mac_address"

This reverts commit 790071347a0a1a89e618eedcd51c687ea783aeb3.

We are seeing kernel panic when enabling two NCSI interfaces at same
time. It looks like the lock in ncsi_rsp_handler_oem_gma() caused the
issue.

Kernel panic log:
```
[  224.323380] 8021q: adding VLAN 0 to HW filter on device eth0
[  224.337533] ftgmac100 1e670000.ethernet eth0: NCSI: Handler for packet type 0x82 returned -19
[  224.358372] BUG: scheduling while atomic: systemd-network/697/0x00000100
[  224.373274] Modules linked in:
[  224.373817] 8021q: adding VLAN 0 to HW filter on device eth1
[  224.380063] CPU: 0 PID: 697 Comm: systemd-network Tainted: G        W          6.6.62-8ea1fc6-dirty-cbd80d0-gcbd80d04d13c #1
[  224.380081] Hardware name: Generic DT based system
[  224.380096]  unwind_backtrace from show_stack+0x18/0x1c
[  224.439407]  show_stack from dump_stack_lvl+0x40/0x4c
[  224.450573]  dump_stack_lvl from __schedule_bug+0x5c/0x70
[  224.462492]  __schedule_bug from __schedule+0x884/0x968
[  224.474026]  __schedule from schedule+0x58/0xa8
[  224.484026]  schedule from schedule_preempt_disabled+0x14/0x18
[  224.496906]  schedule_preempt_disabled from __mutex_lock.constprop.0+0x350/0x76c
[  224.513235]  __mutex_lock.constprop.0 from ncsi_rsp_handler_oem_gma+0x104/0x1a0
[  224.529367]  ncsi_rsp_handler_oem_gma from ncsi_rcv_rsp+0x120/0x2cc
[  224.543195]  ncsi_rcv_rsp from __netif_receive_skb_one_core+0x60/0x84
[  224.557413]  __netif_receive_skb_one_core from netif_receive_skb+0x38/0x148
[  224.572779]  netif_receive_skb from ftgmac100_poll+0x358/0x444
[  224.585656]  ftgmac100_poll from __napi_poll.constprop.0+0x34/0x1d0
[  224.599490]  __napi_poll.constprop.0 from net_rx_action+0x350/0x43c
[  224.613325]  net_rx_action from handle_softirqs+0x114/0x32c
[  224.625624]  handle_softirqs from irq_exit+0x88/0xb8
[  224.636575]  irq_exit from call_with_stack+0x18/0x20
[  224.647530]  call_with_stack from __irq_usr+0x78/0xa0
[  224.658675] Exception stack(0xe075dfb0 to 0xe075dff8)
[  224.669799] dfa0:                                     00000000 00000000 00000000 00000020
[  224.687843] dfc0: 00000069 aefde3e0 00000000 00000000 00000000 00000000 00000000 aefde4e4
[  224.705887] dfe0: 01010101 aefddf20 a6b4331c a6b43618 600f0010 ffffffff
[  224.721100] ------------[ cut here ]------------
[  224.731380] WARNING: CPU: 1 PID: 43 at /lib/list_debug.c:65 __list_del_entry_valid_or_report+0x14c/0x150
[  224.752352] list_del corruption. next->prev should be cfbfdea4, but was 0000000c. (next=cf801de4)
[  224.771953] Modules linked in:
[  224.778715] CPU: 1 PID: 43 Comm: kworker/1:3 Tainted: G        W          6.6.62-8ea1fc6-dirty-cbd80d0-gcbd80d04d13c #1
[  224.802513] Hardware name: Generic DT based system
[  224.813076] Workqueue: ipv6_addrconf addrconf_dad_work
[  224.824428]  unwind_backtrace from show_stack+0x18/0x1c
[  224.835964]  show_stack from dump_stack_lvl+0x40/0x4c
[  224.847113]  dump_stack_lvl from __warn+0x80/0x11c
[  224.857687]  __warn from warn_slowpath_fmt+0x12c/0x194
[  224.869025]  warn_slowpath_fmt from __list_del_entry_valid_or_report+0x14c/0x150
[  224.885352]  __list_del_entry_valid_or_report from __mutex_remove_waiter+0x1c/0x64
[  224.902065]  __mutex_remove_waiter from __mutex_lock.constprop.0+0x160/0x76c
[  224.917624]  __mutex_lock.constprop.0 from addrconf_dad_work+0x48/0x5e0
[  224.932224]  addrconf_dad_work from process_one_work+0x198/0x3c0
[  224.945487]  process_one_work from worker_thread+0x208/0x460
[  224.957978]  worker_thread from kthread+0xe8/0x104
[  224.968552]  kthread from ret_from_fork+0x14/0x28
[  224.978930] Exception stack(0xcfbfdfb0 to 0xcfbfdff8)
[  224.990069] dfa0:                                     00000000 00000000 00000000 00000000
[  225.008112] dfc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[  225.026153] dfe0: 00000000 00000000 00000000 00000000 00000013 00000000
[  225.040816] ---[ end trace 0000000000000000 ]---
[  225.051017] ------------[ cut here ]------------
[  225.061207] WARNING: CPU: 0 PID: 777 at /lib/list_debug.c:32 __list_add_valid_or_report+0xe0/0x10c
[  225.080997] list_add corruption. prev->next should be next (b0f9eed0), but was 00000000. (prev=cf801de4).
[  225.102152] Modules linked in:
[  225.108893] CPU: 0 PID: 777 Comm: phosphor-networ Tainted: G        W          6.6.62-8ea1fc6-dirty-cbd80d0-gcbd80d04d13c #1
[  225.133647] Hardware name: Generic DT based system
[  225.144210]  unwind_backtrace from show_stack+0x18/0x1c
[  225.155743]  show_stack from dump_stack_lvl+0x40/0x4c
[  225.166890]  dump_stack_lvl from __warn+0x80/0x11c
[  225.177463]  __warn from warn_slowpath_fmt+0x12c/0x194
[  225.188799]  warn_slowpath_fmt from __list_add_valid_or_report+0xe0/0x10c
[  225.203782]  __list_add_valid_or_report from __mutex_add_waiter+0x28/0x60
[  225.218768]  __mutex_add_waiter from __mutex_lock.constprop.0+0xe0/0x76c
[  225.233559]  __mutex_lock.constprop.0 from dev_ethtool+0x13c/0x2588
[  225.247396]  dev_ethtool from dev_ioctl+0x3e4/0x648
[  225.258165]  dev_ioctl from sock_ioctl+0x2dc/0x64c
[  225.268741]  sock_ioctl from sys_ioctl+0x364/0xc48
[  225.279313]  sys_ioctl from ret_fast_syscall+0x0/0x54
[  225.290454] Exception stack(0xcf905fa8 to 0xcf905ff0)
[  225.301594] 5fa0:                   00008946 ae8c6178 00000009 00008946 ae8c6178 ae8c6068
[  225.319637] 5fc0: 00008946 ae8c6178 00000004 00000036 0047f2b4 00885ea8 ae8c619d 00000000
[  225.337677] 5fe0: a69c9e20 ae8c6058 a69c04e4 a669e9e4
[  225.348851] ---[ end trace 0000000000000000 ]---
[  225.359044] ------------[ cut here ]------------
[  225.369229] WARNING: CPU: 0 PID: 777 at /lib/list_debug.c:52 __list_del_entry_valid_or_report+0x7c/0x150
[  225.390167] list_del corruption, cf905c34->next is NULL
[  225.401693] Modules linked in:
[  225.408458] CPU: 0 PID: 777 Comm: phosphor-networ Tainted: G        W          6.6.62-8ea1fc6-dirty-cbd80d0-gcbd80d04d13c #1
[  225.433214] Hardware name: Generic DT based system
[  225.443775]  unwind_backtrace from show_stack+0x18/0x1c
[  225.455307]  show_stack from dump_stack_lvl+0x40/0x4c
[  225.466455]  dump_stack_lvl from __warn+0x80/0x11c
[  225.477028]  __warn from warn_slowpath_fmt+0x12c/0x194
[  225.488365]  warn_slowpath_fmt from __list_del_entry_valid_or_report+0x7c/0x150
[  225.504498]  __list_del_entry_valid_or_report from __mutex_remove_waiter+0x1c/0x64
[  225.521209]  __mutex_remove_waiter from __mutex_lock.constprop.0+0x160/0x76c
[  225.536767]  __mutex_lock.constprop.0 from dev_ethtool+0x13c/0x2588
[  225.550599]  dev_ethtool from dev_ioctl+0x3e4/0x648
[  225.561365]  dev_ioctl from sock_ioctl+0x2dc/0x64c
[  225.571939]  sock_ioctl from sys_ioctl+0x364/0xc48
[  225.582509]  sys_ioctl from ret_fast_syscall+0x0/0x54
[  225.593651] Exception stack(0xcf905fa8 to 0xcf905ff0)
[  225.604790] 5fa0:                   00008946 ae8c6178 00000009 00008946 ae8c6178 ae8c6068
[  225.622834] 5fc0: 00008946 ae8c6178 00000004 00000036 0047f2b4 00885ea8 ae8c619d 00000000
[  225.640875] 5fe0: a69c9e20 ae8c6058 a69c04e4 a669e9e4
[  225.652060] ---[ end trace 0000000000000000 ]---
[  225.662248] ------------[ cut here ]------------
[  225.672500] WARNING: CPU: 0 PID: 777 at /lib/refcount.c:22 __mutex_unlock_slowpath+0x98/0x180
[  225.691452] refcount_t: saturated; leaking memory.
[  225.702100] Modules linked in:
[  225.708891] CPU: 0 PID: 777 Comm: phosphor-networ Tainted: G        W          6.6.62-8ea1fc6-dirty-cbd80d0-gcbd80d04d13c #1
[  225.733666] Hardware name: Generic DT based system
[  225.744244]  unwind_backtrace from show_stack+0x18/0x1c
[  225.755799]  show_stack from dump_stack_lvl+0x40/0x4c
[  225.766959]  dump_stack_lvl from __warn+0x80/0x11c
[  225.777536]  __warn from warn_slowpath_fmt+0x12c/0x194
[  225.788873]  warn_slowpath_fmt from __mutex_unlock_slowpath+0x98/0x180
[  225.803278]  __mutex_unlock_slowpath from __rtnl_unlock+0x40/0x80
[  225.816729]  __rtnl_unlock from netdev_run_todo+0x54/0x5cc
[  225.828840]  netdev_run_todo from dev_ethtool+0x4d0/0x2588
[  225.840952]  dev_ethtool from dev_ioctl+0x3e4/0x648
[  225.851721]  dev_ioctl from sock_ioctl+0x2dc/0x64c
[  225.862298]  sock_ioctl from sys_ioctl+0x364/0xc48
[  225.872871]  sys_ioctl from ret_fast_syscall+0x0/0x54
[  225.884014] Exception stack(0xcf905fa8 to 0xcf905ff0)
[  225.895156] 5fa0:                   00008946 ae8c6178 00000009 00008946 ae8c6178 ae8c6068
[  225.913201] 5fc0: 00008946 ae8c6178 00000004 00000036 0047f2b4 00885ea8 ae8c619d 00000000
[  225.931243] 5fe0: a69c9e20 ae8c6058 a69c04e4 a669e9e4
[  225.942450] ---[ end trace 0000000000000000 ]---
[  225.952669] ------------[ cut here ]------------
[  225.952670] ------------[ cut here ]------------
[  225.952676] WARNING: CPU: 1 PID: 43 at /lib/refcount.c:25 __mutex_unlock_slowpath+0x98/0x180
[  225.962857] WARNING: CPU: 0 PID: 777 at /lib/refcount.c:28 wake_up_q+0x9c/0xa8
[  225.973037] refcount_t: addition on 0; use-after-free.
[  225.973041] Modules linked in:
[  225.991653] refcount_t: underflow; use-after-free.
[  225.991657] Modules linked in:
[  226.007590]
[  226.007598] CPU: 1 PID: 43 Comm: kworker/1:3 Tainted: G        W          6.6.62-8ea1fc6-dirty-cbd80d0-gcbd80d04d13c #1
[  226.007608] Hardware name: Generic DT based system
[  226.007613] Workqueue: ipv6_addrconf addrconf_dad_work
[  226.018949]
[  226.025657]
[  226.025667]  unwind_backtrace from show_stack+0x18/0x1c
[  226.109926]  show_stack from dump_stack_lvl+0x40/0x4c
[  226.121073]  dump_stack_lvl from __warn+0x80/0x11c
[  226.131645]  __warn from warn_slowpath_fmt+0x12c/0x194
[  226.142983]  warn_slowpath_fmt from __mutex_unlock_slowpath+0x98/0x180
[  226.157386]  __mutex_unlock_slowpath from __rtnl_unlock+0x40/0x80
[  226.170833]  __rtnl_unlock from netdev_run_todo+0x54/0x5cc
[  226.182941]  netdev_run_todo from addrconf_dad_work+0x1dc/0x5e0
[  226.196008]  addrconf_dad_work from process_one_work+0x198/0x3c0
[  226.209270]  process_one_work from worker_thread+0x208/0x460
[  226.221763]  worker_thread from kthread+0xe8/0x104
[  226.232337]  kthread from ret_from_fork+0x14/0x28
[  226.242714] Exception stack(0xcfbfdfb0 to 0xcfbfdff8)
[  226.253853] dfa0:                                     00000000 00000000 00000000 00000000
[  226.271894] dfc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[  226.289936] dfe0: 00000000 00000000 00000000 00000000 00000013 00000000
[  226.304524] CPU: 0 PID: 777 Comm: phosphor-networ Tainted: G        W          6.6.62-8ea1fc6-dirty-cbd80d0-gcbd80d04d13c #1
[  226.304594] ---[ end trace 0000000000000000 ]---
[  226.329278] Hardware name: Generic DT based system
[  226.329285]  unwind_backtrace from show_stack+0x18/0x1c
[  226.361555]  show_stack from dump_stack_lvl+0x40/0x4c
[  226.372725]  dump_stack_lvl from __warn+0x80/0x11c
[  226.381207] 8<--- cut here ---
[  226.383304]  __warn from warn_slowpath_fmt+0x12c/0x194
[  226.390113] Unable to handle kernel NULL pointer dereference at virtual address 00000790 when read
[  226.401357]  warn_slowpath_fmt from wake_up_q+0x9c/0xa8
[  226.401375]  wake_up_q from __mutex_unlock_slowpath+0xb8/0x180
[  226.401389]  __mutex_unlock_slowpath from __rtnl_unlock+0x40/0x80
[  226.401407]  __rtnl_unlock from netdev_run_todo+0x54/0x5cc
[  226.401427]  netdev_run_todo from dev_ethtool+0x4d0/0x2588
[  226.421212] [00000790] *pgd=00000000
[  226.432699]  dev_ethtool from dev_ioctl+0x3e4/0x648
[  226.432722]  dev_ioctl from sock_ioctl+0x2dc/0x64c
[  226.445601]
[  226.459015]  sock_ioctl from sys_ioctl+0x364/0xc48
[  226.459034]  sys_ioctl from ret_fast_syscall+0x0/0x54
[  226.471136] Internal error: Oops: 5 [#1] SMP ARM
[  226.483207] Exception stack(0xcf905fa8 to 0xcf905ff0)
[  226.491077] Modules linked in:
[  226.501822] 5fa0:                   00008946 ae8c6178 00000009 00008946 ae8c6178 ae8c6068
[  226.512368]
[  226.515637] 5fc0: 00008946 ae8c6178 00000004 00000036 0047f2b4 00885ea8 ae8c619d 00000000
[  226.526185] CPU: 1 PID: 43 Comm: kworker/1:3 Tainted: G        W          6.6.62-8ea1fc6-dirty-cbd80d0-gcbd80d04d13c #1
[  226.537312] 5fe0: a69c9e20 ae8c6058 a69c04e4 a669e9e4
[  226.547477] Hardware name: Generic DT based system
[  226.558716] ---[ end trace 0000000000000000 ]---
[  226.565321] Workqueue: ipv6_addrconf addrconf_dad_work
[  226.565346] PC is at wake_q_add+0x20/0x98
[  226.565360] LR is at __mutex_unlock_slowpath+0x98/0x180
[  226.608385] ------------[ cut here ]------------
[  226.628480] pc : [<b015537c>]    lr : [<b0a2d428>]    psr: 200f0013
[  226.628487] sp : cfbfde58  ip : 00000000  fp : cfbfde5c
[  226.628492] r10: b117f500  r9 : 0000afa1  r8 : 3b807e28
[  226.639628] WARNING: CPU: 0 PID: 0 at /kernel/context_tracking.c:128 ct_kernel_exit.constprop.0+0x110/0x11c
[  226.650177] r7 : bfdbf780  r6 : cfbfde64  r5 : 00000000  r4 : b0f9eec4
[  226.660347] Modules linked in:
[  226.671664] r3 : 00000790  r2 : 00000001  r1 : 00000000  r0 : cfbfde64
[  226.680489]
[  226.691999] Flags: nzCv  IRQs on  FIQs on  Mode SVC_32  ISA ARM  Segment none
[  226.702171] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G        W          6.6.62-8ea1fc6-dirty-cbd80d0-gcbd80d04d13c #1
[  226.715979] Control: 10c5387d  Table: 8f42006a  DAC: 00000051
[  226.727488] Hardware name: Generic DT based system
[  226.738997] Register r0 information:
[  226.760487]  unwind_backtrace from show_stack+0x18/0x1c
[  226.774865]  2-page vmalloc region starting at 0xcfbfc000 allocated at kernel_clone+0xac/0x37c
[  226.781580]  show_stack from dump_stack_lvl+0x40/0x4c
[  226.795966] Register r1 information:
[  226.799232]  dump_stack_lvl from __warn+0x80/0x11c
[  226.814960]  NULL pointer
[  226.838166]  __warn from warn_slowpath_fmt+0x18c/0x194
[  226.850825] Register r2 information:
[  226.861378]  warn_slowpath_fmt from ct_kernel_exit.constprop.0+0x110/0x11c
[  226.869243]  non-paged memory
[  226.880754]  ct_kernel_exit.constprop.0 from default_idle_call+0x24/0xa0
[  226.899740] Register r3 information:
[  226.910867]  default_idle_call from do_idle+0xac/0x120
[  226.918733]  non-paged memory
[  226.929285]  do_idle from cpu_startup_entry+0x30/0x34
[  226.935040] Register r4 information:
[  226.946358]  cpu_startup_entry from rest_init+0xb4/0xb8
[  226.954226]  non-slab/vmalloc memory
[  226.969379]  rest_init from arch_post_acpi_subsys_init+0x0/0x18
[  226.975902] Register r5 information: NULL pointer
[  226.990676] ---[ end trace 0000000000000000 ]---
[  226.998538] Register r6 information: 2-page vmalloc region starting at 0xcfbfc000 allocated at kernel_clone+0xac/0x37c
[  227.112139] Register r7 information: slab kmalloc-192 start bfdbf780 pointer offset 0 size 192
[  227.131164] Register r8 information: non-paged memory
[  227.142304] Register r9 information: non-paged memory
[  227.153426] Register r10 information: slab task_struct start b117f500 pointer offset 0 size 2496
[  227.172826] Register r11 information: 2-page vmalloc region starting at 0xcfbfc000 allocated at kernel_clone+0xac/0x37c
[  227.196630] Register r12 information: NULL pointer
[  227.207193] Process kworker/1:3 (pid: 43, stack limit = 0x5e87f8b8)
[  227.221019] Stack: (0xcfbfde58 to 0xcfbfe000)
[  227.230624] de40:                                                       ce7e1f40 b0a2d428
[  227.248669] de60: 00000000 00000001 cfbfde64 ecefe4cd 00000000 bfdbf7a4 cfbfdea4 b08a7274
[  227.266713] de80: bfdbf7a0 b0891ea0 cfbfdee8 3b807e28 0000afa1 b117f500 ce7e1f40 b09b6cf0
[  227.284756] dea0: 3b807e28 cfbfdea4 cfbfdea4 00000000 00000000 ecefe4cd 00000000 bfdbf7a0
[  227.302800] dec0: bfdbf7a4 bf8f7800 bfdbf780 3b807e28 0000afa1 b117f500 ce7e1f40 b099dcfc
[  227.320843] dee0: 3b807e28 0000afa1 000002ff 00000000 01000000 812eb1ff 000002ff b117f500
[  227.338886] df00: bfdbf348 b1f79400 ce7e1f40 ecefe4cd b1f79405 b1264b80 bfdbf7c8 b1f79400
[  227.356929] df20: ce7e1f40 00000040 b1f79405 b013e270 61c88647 b1264bac b117f500 b1264b80
[  227.374973] df40: ce7e1f40 b0f03d40 ce7e1f60 61c88647 b1264bac b117f500 ce7e1f40 b013e6a0
[  227.393016] df60: ce7e1f9c b0f03d40 cfbf5ed0 b182ec80 b117f500 b013e498 b1264b80 b182ed40
[  227.411058] df80: cfbf5ed0 00000000 00000000 b0147ac8 b182ec80 b01479e0 00000000 00000000
[  227.429101] dfa0: 00000000 00000000 00000000 b010014c 00000000 00000000 00000000 00000000
[  227.447144] dfc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[  227.465186] dfe0: 00000000 00000000 00000000 00000000 00000013 00000000 00000000 00000000
[  227.483233]  wake_q_add from __mutex_unlock_slowpath+0x98/0x180
[  227.496306]  __mutex_unlock_slowpath from __rtnl_unlock+0x40/0x80
[  227.509756]  __rtnl_unlock from netdev_run_todo+0x54/0x5cc
[  227.521866]  netdev_run_todo from addrconf_dad_work+0x1dc/0x5e0
[  227.534937]  addrconf_dad_work from process_one_work+0x198/0x3c0
[  227.548202]  process_one_work from worker_thread+0x208/0x460
[  227.560695]  worker_thread from kthread+0xe8/0x104
[  227.571269]  kthread from ret_from_fork+0x14/0x28
[  227.581648] Exception stack(0xcfbfdfb0 to 0xcfbfdff8)
[  227.592786] dfa0:                                     00000000 00000000 00000000 00000000
[  227.610811] dfc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[  227.628854] dfe0: 00000000 00000000 00000000 00000000 00000013 00000000
[  227.643447] Code: e2813e79 f57ff05b f593f000 e3a02001 (e193ef9f)
[  227.657073] ---[ end trace 0000000000000000 ]---
[  227.667344] Kernel panic - not syncing: Fatal exception
[  227.678889] CPU0: stopping
[  227.678909] CPU: 0 PID: 1620 Comm: journal-offline Tainted: G      D W          6.6.62-8ea1fc6-dirty-cbd80d0-gcbd80d04d13c #1
[  227.678922] Hardware name: Generic DT based system
[  227.678935]  unwind_backtrace from show_stack+0x18/0x1c
[  227.678964]  show_stack from dump_stack_lvl+0x40/0x4c
[  227.678983]  dump_stack_lvl from do_handle_IPI+0x18c/0x1c0
[  227.678999]  do_handle_IPI from ipi_handler+0x20/0x28
[  227.679012]  ipi_handler from handle_percpu_devid_irq+0x84/0x1c0
[  227.679027]  handle_percpu_devid_irq from generic_handle_domain_irq+0x30/0x40
[  227.679042]  generic_handle_domain_irq from gic_handle_irq+0x6c/0x80
[  227.679058]  gic_handle_irq from generic_handle_arch_irq+0x34/0x44
[  227.679074]  generic_handle_arch_irq from call_with_stack+0x18/0x20
[  227.679092]  call_with_stack from __irq_svc+0x98/0xb0
[  227.679104] Exception stack(0xcf9757a8 to 0xcf9757f0)
[  227.679114] 57a0:                   cfe80000 b1862040 00000000 00000000 00000001 b1306b40
[  227.679122] 57c0: b1306b40 00000000 00002a33 b1306b60 b1c7b040 00002a33 00000000 cf9757f8
[  227.679129] 57e0: b06fc124 b06fc12c 40070013 ffffffff
[  227.679134]  __irq_svc from aspeed_spi_exec_op+0x410/0x440
[  227.679150]  aspeed_spi_exec_op from spi_mem_exec_op+0x3f8/0x46c
[  227.679162]  spi_mem_exec_op from spi_nor_read_sr+0xb0/0x138
[  227.679175]  spi_nor_read_sr from spi_nor_wait_till_ready_with_timeout+0xd0/0x19c
[  227.679186]  spi_nor_wait_till_ready_with_timeout from spi_nor_write+0x148/0x298
[  227.679197]  spi_nor_write from mtd_write_oob_std+0xa0/0xa8
[  227.679210]  mtd_write_oob_std from mtd_write+0x68/0x94
[  227.679225]  mtd_write from ubi_io_write+0x120/0x68c
[  227.679245]  ubi_io_write from ubi_eba_write_leb+0xe4/0x7a4
[  227.679265]  ubi_eba_write_leb from ubi_leb_write+0xe4/0x12c
[  227.679283]  ubi_leb_write from ubifs_leb_write+0xa8/0x10c
[  227.679302]  ubifs_leb_write from ubifs_wbuf_write_nolock+0x2b8/0xa48
[  227.679315]  ubifs_wbuf_write_nolock from write_head+0x70/0xd8
[  227.679335]  write_head from ubifs_jnl_write_data+0x224/0x2f0
[  227.679355]  ubifs_jnl_write_data from do_writepage+0x224/0x2dc
[  227.679370]  do_writepage from writepage_cb+0x1c/0x70
[  227.679386]  writepage_cb from write_cache_pages+0x148/0x3cc
[  227.679404]  write_cache_pages from do_writepages+0xf4/0x1cc
[  227.679421]  do_writepages from __filemap_fdatawrite_range+0xa8/0xb0
[  227.679437]  __filemap_fdatawrite_range from file_write_and_wait_range+0x64/0xc8
[  227.679449]  file_write_and_wait_range from ubifs_fsync+0x94/0x108
[  227.679465]  ubifs_fsync from ovl_fsync+0xf4/0x108
[  227.679482]  ovl_fsync from do_fsync+0x48/0x7c
[  227.679495]  do_fsync from __sys_trace_return+0x0/0x10
[  227.679506] Exception stack(0xcf975fa8 to 0xcf975ff0)
[  227.679514] 5fa0:                   00000010 00d28248 00000010 00000002 a663f840 00000000
[  227.679523] 5fc0: 00000010 00d28248 a68c3600 00000076 00000000 a663f340 a5e3f000 a663ee3c
[  227.679529] 5fe0: a663f3c0 a663ec58 00000000 a6bbbcfc
```
---
 net/ncsi/ncsi-rsp.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/net/ncsi/ncsi-rsp.c b/net/ncsi/ncsi-rsp.c
index 5130e30ebcae2..b51e8adc68e76 100644
--- a/net/ncsi/ncsi-rsp.c
+++ b/net/ncsi/ncsi-rsp.c
@@ -629,6 +629,7 @@ static int ncsi_rsp_handler_oem_gma(struct ncsi_request *nr, int mfr_id)
 {
 	struct ncsi_dev_priv *ndp = nr->ndp;
 	struct net_device *ndev = ndp->ndev.dev;
+	const struct net_device_ops *ops = ndev->netdev_ops;
 	struct ncsi_rsp_oem_pkt *rsp;
 	struct sockaddr saddr;
 	u32 mac_addr_off = 0;
@@ -655,9 +656,7 @@ static int ncsi_rsp_handler_oem_gma(struct ncsi_request *nr, int mfr_id)
 	/* Set the flag for GMA command which should only be called once */
 	ndp->gma_flag = 1;
 
-	rtnl_lock();
-	ret = dev_set_mac_address(ndev, &saddr, NULL);
-	rtnl_unlock();
+	ret = ops->ndo_set_mac_address(ndev, &saddr);
 	if (ret < 0)
 		netdev_warn(ndev, "NCSI: 'Writing mac address to device failed\n");
 
-- 
2.31.1

